File: node/deliberation/src/specialists/crypto-majors.mjs
/**
 * @accretive-substrate/deliberation/specialists/crypto-majors
 *
 * Crypto-majors specialist. Reads venue-specific market state for major
 * crypto symbols (BTC, ETH, SOL, …) and emits verdict based on:
 *   - Funding rate (Binance futures) — flat-to-positive = healthy basis
 *   - Order book depth (KuCoin L2_20) — thicker = supports verdict
 *   - 24h trend direction + magnitude
 *
 * Verdict rules:
 *   funding > 0.05% AND 24h_chg > 1%   → buy   (healthy basis + trend up)
 *   funding < -0.03% OR 24h_chg < -3% → sell  (basis flipping bear OR sharp dump)
 *   else                              → hold  (no clear edge)
 *
 * Domain: crypto-majors (kucoin/binance + symbol in MAJORS list).
 */

import { makeSpecialist, voteFor, abstainBecause, matchesDomain } from './base.mjs';

export const CRYPTO_MAJORS_SPECIALIST_NAME = 'crypto-majors';
const MAJORS = new Set(['BTC', 'ETH', 'SOL', 'BNB', 'XRP', 'ADA', 'AVAX', 'DOT', 'LINK']);

async function argue(pkg) {
  const venue = String(pkg?.trigger?.venue || '').toLowerCase();
  if (!['kucoin', 'binance'].includes(venue)) {
    return abstainBecause(CRYPTO_MAJORS_SPECIALIST_NAME, `venue=${venue} out of scope`);
  }

  const symbol = String(pkg?.trigger?.symbol || '');
  const base = symbol.split(/[-/]/)[0]?.toUpperCase();
  if (!MAJORS.has(base)) {
    return abstainBecause(CRYPTO_MAJORS_SPECIALIST_NAME, `symbol ${symbol} not a major`);
  }

  const market = pkg?.market_state?.[symbol] || pkg?.market_state || {};
  const fundingRate = Number(market.funding_rate);
  const chg24 = Number(market.change_24h_pct);
  const orderbookDepth = Number(market.orderbook_depth_usd);

  const citations = [
    { type: 'signal', source: 'crypto-majors', symbol, market: { funding_rate: fundingRate, change_24h_pct: chg24 } },
  ];

  if (!Number.isFinite(fundingRate) && !Number.isFinite(chg24)) {
    return abstainBecause(CRYPTO_MAJORS_SPECIALIST_NAME, 'no funding/24h data in evidence');
  }

  // Healthy basis + trend up.
  if (fundingRate > 0.0005 && chg24 > 1) {
    const conf = Math.min(0.85, 0.6 + chg24 / 20 + fundingRate * 100);
    return voteFor(
      CRYPTO_MAJORS_SPECIALIST_NAME, 'buy', conf,
      `funding ${(fundingRate*100).toFixed(3)}% + 24h ${chg24.toFixed(1)}%${Number.isFinite(orderbookDepth) ? ` + depth $${Math.round(orderbookDepth).toLocaleString()}` : ''}`,
      citations,
    );
  }

  // Bear basis or sharp dump.
  if (fundingRate < -0.0003 || chg24 < -3) {
    const reason = fundingRate < -0.0003
      ? `funding flipped negative ${(fundingRate*100).toFixed(3)}%`
      : `24h drop ${chg24.toFixed(1)}% > -3% threshold`;
    const conf = Math.min(0.85, 0.6 + Math.abs(chg24) / 25);
    return voteFor(CRYPTO_MAJORS_SPECIALIST_NAME, 'sell', conf, reason, citations);
  }

  return voteFor(
    CRYPTO_MAJORS_SPECIALIST_NAME, 'hold', 0.5,
    `funding ${Number.isFinite(fundingRate) ? (fundingRate*100).toFixed(3)+'%' : 'n/a'} 24h ${Number.isFinite(chg24) ? chg24.toFixed(1)+'%' : 'n/a'} — no clear edge`,
    citations,
  );
}

export const cryptoMajorsSpecialist = makeSpecialist(
  CRYPTO_MAJORS_SPECIALIST_NAME,
  ['crypto'],
  argue,
);

File: go/deliberation/pkg/deliberation/specialists/cryptomajors/cryptomajors.go
// Package cryptomajors — Go port of @accretive-substrate/deliberation/specialists/crypto-majors.
//
// Filters venue=kucoin|binance + symbol base in MAJORS list. Reads funding
// rate + 24h chg + (optional) orderbook depth from MarketState.
//
// Verdict rules (mirror Node):
//   funding>0.05% AND chg24>1%   → buy
//   funding<-0.03% OR chg24<-3%  → sell
//   else                          → hold
package cryptomajors

import (
	"fmt"
	"math"
	"strings"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation"
)

const Name = "crypto-majors"

var majors = map[string]bool{
	"BTC": true, "ETH": true, "SOL": true, "BNB": true, "XRP": true,
	"ADA": true, "AVAX": true, "DOT": true, "LINK": true,
}

type majorsMarket struct {
	FundingRate       float64 `json:"funding_rate"`
	Change24hPct      float64 `json:"change_24h_pct"`
	OrderbookDepthUSD float64 `json:"orderbook_depth_usd"`
}

func Argue(pkg deliberation.EvidencePackage) deliberation.Voice {
	venue := strings.ToLower(pkg.Trigger.Venue)
	if venue != "kucoin" && venue != "binance" {
		return deliberation.AbstainBecause(Name, fmt.Sprintf("venue=%s out of scope", venue))
	}

	symbol := pkg.Trigger.Symbol
	base := strings.ToUpper(strings.Split(symbol, "-")[0])
	if !majors[base] {
		return deliberation.AbstainBecause(Name, fmt.Sprintf("symbol %s not a major", symbol))
	}

	var market majorsMarket
	_ = deliberation.DecodeMarketStateFor(pkg.MarketState, symbol, &market)

	citations := []deliberation.Citation{{
		Type:   "signal",
		Source: "crypto-majors",
		Symbol: symbol,
	}}

	// No data at all.
	if market.FundingRate == 0 && market.Change24hPct == 0 {
		return deliberation.AbstainBecause(Name, "no funding/24h data in evidence")
	}

	// Healthy basis + trend up → buy.
	if market.FundingRate > 0.0005 && market.Change24hPct > 1 {
		conf := math.Min(0.85, 0.6+market.Change24hPct/20+market.FundingRate*100)
		depth := ""
		if market.OrderbookDepthUSD > 0 {
			depth = fmt.Sprintf(" + depth $%.0f", market.OrderbookDepthUSD)
		}
		return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf,
			fmt.Sprintf("funding %.3f%% + 24h %.1f%%%s",
				market.FundingRate*100, market.Change24hPct, depth),
			citations)
	}

	// Bear basis or sharp dump → sell.
	if market.FundingRate < -0.0003 || market.Change24hPct < -3 {
		reason := fmt.Sprintf("24h drop %.1f%% > -3%% threshold", market.Change24hPct)
		if market.FundingRate < -0.0003 {
			reason = fmt.Sprintf("funding flipped negative %.3f%%", market.FundingRate*100)
		}
		conf := math.Min(0.85, 0.6+math.Abs(market.Change24hPct)/25)
		return deliberation.VoteFor(Name, deliberation.VerdictSell, conf, reason, citations)
	}

	return deliberation.VoteFor(Name, deliberation.VerdictHold, 0.5,
		fmt.Sprintf("funding %.3f%% 24h %.1f%% — no clear edge",
			market.FundingRate*100, market.Change24hPct),
		citations)
}

// Specialist — exported safe-wrapped argue function.
var Specialist = deliberation.MakeSafe(Name, Argue)
